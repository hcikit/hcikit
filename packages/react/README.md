# @hcikit/react

This package is a package that allows you to create and run HCI experiments. Visit hcikit.github.io for more information.
