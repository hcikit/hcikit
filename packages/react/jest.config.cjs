/** @type {import('@ts-jest/dist/types').InitialOptionsTsJest} */
module.exports = {
  preset: "ts-jest",
  testEnvironment: "jsdom",
  moduleNameMapper: {
    "@hcikit/workflow(.*)$": "<rootDir>/../workflow/src$1",
  },
  testMatch: ["./src/**/*.[jt]s?(x)", "**/?(*.)+(spec|test).[t]s?(x)"],
  modulePathIgnorePatterns: ["./dist/", "./node_modules/"],
  testPathIgnorePatterns: ["./dist/", "./node_modules/"],
  transformIgnorePatterns: ["./dist/", "/node_modules/"],
  setupFilesAfterEnv: ["<rootDir>/setupTests.ts"],
  collectCoverageFrom: [
    "src/**/*.{ts,tsx}",
    "!/node_modules/",
    "!src/setupTests.ts",
    "!src/**/*.stories.{js,ts,tsx,jsx}",
  ],
};
