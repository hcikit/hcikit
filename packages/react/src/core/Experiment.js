import { throttle } from "lodash-es";
import React, { useEffect, useState } from "react";
import { Provider } from "react-redux";
import PropTypes from "prop-types";

import { configureStore } from "@hcikit/workflow";
import TaskRenderer from "./TaskRenderer";

const STATE_KEY = "HCIKIT_LOGS";

export const loadStateFromSessionStorage = () => {
  try {
    const state = window.sessionStorage.getItem(STATE_KEY);
    if (state) {
      return JSON.parse(state);
    }
  } catch (err) {
    console.error("Failed to load from sessionStorage", err);
  }

  return undefined;
};

export const saveStateToSessionStorage = throttle((state) => {
  try {
    window.sessionStorage.setItem(STATE_KEY, JSON.stringify(state));
  } catch (err) {
    console.error("Failed to save to sessionStorage", err);
  }
}, 3000);

let Experiment = ({
  saveState = saveStateToSessionStorage,
  loadState = loadStateFromSessionStorage,
  ...props
}) => {
  // TODO: maybe we should hide the task registry and instead just pass a list of objects?
  // TODO: not sure how to create different sessions for the same task.

  let [store, setStore] = useState();
  console.log(setStore);
  console.log({ store });
  /* eslint-disable react-hooks/exhaustive-deps */
  useEffect(() => {
    let storedState;

    if (process.env.NODE_ENV !== "development" && loadState) {
      storedState = loadState();
    }

    setStore(
      configureStore({ ...props.configuration, ...storedState }, saveState)
    );
  }, []);

  if (!store) {
    return null;
  }

  return (
    <Provider store={store}>
      <TaskRenderer {...props} />
    </Provider>
  );
};

Experiment.propTypes = {
  configuration: PropTypes.object.isRequired,
  loadState: PropTypes.func,
  saveState: PropTypes.func,
  tasks: PropTypes.objectOf(PropTypes.elementType),
};

export default Experiment;
