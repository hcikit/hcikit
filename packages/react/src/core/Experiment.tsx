import { isEmpty } from "lodash-es";
import React, {
  useState,
  createContext,
  useContext,
  useCallback,
  ElementType,
  useMemo,
  useEffect,
} from "react";

import { serializeError } from "serialize-error";
import TaskRenderer from "./TaskRenderer.js";

import {
  advanceConfiguration,
  logToConfiguration,
  modifyConfiguration,
  Configuration,
  ExperimentIndex,
  UnfilledLog,
  experimentComplete,
  getCurrentIndex,
  TypedConfiguration,
} from "@hcikit/workflow";

import GridLayout from "../GridLayout.js";

import { ErrorBoundary, FallbackProps } from "react-error-boundary";
import { CenteredNicePaper } from "../components/index.js";
import { BasePersistence, StoragePersistence } from "../persistence/index.js";

import { Button, Typography, CircularProgress } from "@mui/material";
import { Task } from "../index.js";

type PropsOfDict<T extends Record<string, React.ComponentType<any>>> = {
  [Task in keyof T]: React.ComponentProps<T[Task]>;
};

export type ConfigurationReact<
  T extends Record<string, React.ComponentType<any>>,
> = TypedConfiguration<PropsOfDict<T>>;

export interface ControlFunctions {
  advance: (index?: ExperimentIndex) => void;
  log: (log: UnfilledLog) => void;
  modify: (
    modifiedConfig: Record<string, unknown>,
    index?: ExperimentIndex
  ) => void;
  persistence?: BasePersistence | null;
}

export const ConfigContext = createContext<Configuration | undefined>(
  undefined
);

export const ConfigMutatorContext = createContext<ControlFunctions | undefined>(
  undefined
);

function useConfiguration(): Configuration {
  const context = useContext(ConfigContext);
  if (context === undefined) {
    throw new Error("useConfig must be used within a ConfigProvider");
  }
  return context;
}

function useExperiment(): ControlFunctions {
  const context = useContext(ConfigMutatorContext);

  if (context === undefined) {
    throw new Error("useConfig must be used within a ConfigMutatorContext");
  }

  return context;
}

export { useConfiguration, useExperiment };

function isPromise<T extends object>(obj: T | Promise<T>): obj is Promise<T> {
  return "then" in obj && typeof obj.then === "function";
}

const Experiment: React.FunctionComponent<{
  persistence?: BasePersistence | null;
  configuration: Configuration | Promise<Configuration>;
  tasks: Record<string, Task<any>>;
  Layout?: ElementType;
  ErrorHandler?: React.ComponentType<FallbackProps>;
  forceRemountEveryTask?: boolean;
  devOptions?: { persistInDevelopment: boolean; useIndexFromUrl: boolean };
}> = ({
  persistence = new StoragePersistence(window.sessionStorage, "HCIKIT_LOGS"),
  configuration: initialConfiguration,
  devOptions = { persistInDevelopment: false, useIndexFromUrl: true },
  ...props
}) => {
  const shouldPersistConfiguration =
    (process.env.NODE_ENV !== "development" ||
      devOptions.persistInDevelopment) &&
    persistence !== null;

  const [configurations, setConfigurations] = useState<{
    loadedConfiguration: Configuration | undefined;
    initialConfiguration: Configuration | undefined;
    doneLoadingPersisted: boolean;
    doneLoadingInitial: boolean;
  }>({
    loadedConfiguration: undefined,
    initialConfiguration: isPromise(initialConfiguration)
      ? undefined
      : initialConfiguration,
    doneLoadingPersisted: !shouldPersistConfiguration,
    doneLoadingInitial: !isPromise(initialConfiguration),
  });

  useEffect(() => {
    async function loadConfigs() {
      if (shouldPersistConfiguration) {
        persistence?.load().then((loadedConfig) => {
          setConfigurations((c) => ({
            ...c,
            loadedConfiguration: loadedConfig,
            doneLoadingPersisted: true,
          }));
        });
      }
      if (isPromise(initialConfiguration)) {
        initialConfiguration.then((loadedConfig) => {
          setConfigurations((c) => ({
            ...c,
            initialConfiguration: loadedConfig,
            doneLoadingInitial: true,
          }));
        });
      }
    }

    loadConfigs();
  }, []);

  if (
    !configurations.doneLoadingInitial ||
    !configurations.doneLoadingPersisted ||
    !configurations.initialConfiguration
  ) {
    return (
      <CenteredNicePaper>
        <Typography>Loading experiment configuration...</Typography>
        <br />
        <CircularProgress size={100} />
      </CenteredNicePaper>
    );
  } else {
    return (
      <ExperimentInner
        {...props}
        devOptions={devOptions}
        initialConfiguration={configurations.initialConfiguration}
        loadedConfiguration={configurations.loadedConfiguration}
        persistence={persistence}
      />
    );
  }
};

const ExperimentInner: React.FC<{
  persistence?: BasePersistence | null;
  initialConfiguration: Configuration;
  loadedConfiguration: Configuration | undefined;
  tasks: Record<string, Task>;
  Layout?: ElementType;
  ErrorHandler?: React.ComponentType<FallbackProps>;
  forceRemountEveryTask?: boolean;
  devOptions: { persistInDevelopment: boolean; useIndexFromUrl: boolean };
}> = ({
  persistence,
  tasks,
  Layout = GridLayout,
  ErrorHandler = DefaultErrorHandler,
  forceRemountEveryTask,
  initialConfiguration,
  loadedConfiguration,
  devOptions,
}) => {
  // TODO: not sure how to create different sessions for the same task. The issue is that they'll be overwritten by the other thing. Maybe we can add a config version or session key or something to it?
  // TODO: using config and configuration is so confusing...
  const [config, setConfig] = useState<Configuration>(() => {
    let configurationToUse: Configuration = initialConfiguration;

    if (
      loadedConfiguration &&
      (process.env.NODE_ENV !== "development" ||
        devOptions.persistInDevelopment)
    ) {
      configurationToUse = loadedConfiguration;
    }

    // This seems hacky. I think I am assigning the initial index to whatever it should be if the experiment just started, or assigning it to the proper one if we haven't started yet.
    configurationToUse = advanceConfiguration(
      configurationToUse,
      getCurrentIndex(configurationToUse)
    );

    return configurationToUse;
  });

  // it might be better to do this as we save state rather than waiting a render, I chose this way because it is easier to implement than adding it to each of the modifying functions below.
  // useEffect(() => {
  //   if (saveState) {
  //     saveState(config);
  //     console.log("saving");
  //   }
  // }, [config, saveState]);

  const advance = useCallback(
    (index?: ExperimentIndex) =>
      setConfig((c) => {
        c = logToConfiguration(c, { type: "END" });
        c = advanceConfiguration(c, index);

        if (!experimentComplete(c)) {
          c = logToConfiguration(c, { type: "START" });
        }

        persistence?.saveThrottled(c);

        return c;
      }),
    [persistence]
  );

  const log = useCallback(
    (log: UnfilledLog): void =>
      setConfig((c) => {
        const newConfig: Configuration = logToConfiguration(c, log);

        persistence?.saveThrottled(c);

        return newConfig;
      }),
    [persistence]
  );

  const modify = useCallback(
    (modifiedConfig: Record<string, unknown>, index?: ExperimentIndex): void =>
      setConfig((c) => {
        const newConfig = modifyConfiguration(c, modifiedConfig, index);

        persistence?.saveThrottled(c);

        return newConfig;
      }),
    [persistence]
  );

  // Kent says this is actually slower than just recreating it... But honestly idk if I trust that. Truth is I should test it, but this does solve the problem of unnecessary rerenders.

  const experiment = useMemo(
    () => ({
      advance: advance,
      log,
      modify,
      persistence,
    }),
    [advance, log, modify]
  );

  useEffect(() => {
    experiment.log({ type: "START" });
    // This is on purpose, we actually never want to rerun this. But also experiment should never actually change.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    window.onbeforeunload = (e) => {
      persistence?.save(config);
      persistence?.flush();
      return "You have some unsaved changes";
    };
  }, []);

  if (isEmpty(config)) {
    return null;
  }

  return (
    <ConfigMutatorContext.Provider value={experiment}>
      <ConfigContext.Provider value={config}>
        <ErrorBoundary
          onError={(error, info) =>
            log({
              type: "ERROR",
              ...serializeError(info),
              ...serializeError(error),
            })
          }
          FallbackComponent={ErrorHandler}
        >
          <Layout>
            {!experimentComplete(config) ? (
              <TaskRenderer
                tasks={tasks}
                forceRemountEveryTask={forceRemountEveryTask}
              />
            ) : (
              <div style={{ gridArea: "task" }}>
                <CenteredNicePaper>
                  <h2>You&apos;ve completed the experiment!</h2>
                  <a
                    download={`${config.participant || "log"}.json`}
                    href={`data:text/json;charset=utf-8,${encodeURIComponent(
                      JSON.stringify(config)
                    )}`}
                  >
                    Download experiment log
                  </a>
                </CenteredNicePaper>
              </div>
            )}
          </Layout>
        </ErrorBoundary>
      </ConfigContext.Provider>
    </ConfigMutatorContext.Provider>
  );
};

const DefaultErrorHandler: React.FunctionComponent<FallbackProps> = ({
  error,
}) => {
  const configuration = useConfiguration();
  return (
    <CenteredNicePaper>
      <div style={{ gridArea: "task" }}>
        <h2>An error occurred in the experiment.</h2>
        <p style={{ fontStyle: "italic" }}>
          ({error.name}: {error.message})
        </p>
        <br />

        <a
          download={`${configuration.participant || "log"}.json`}
          href={`data:text/json;charset=utf-8,${encodeURIComponent(
            JSON.stringify(configuration)
          )}`}
        >
          Download experiment log
        </a>
      </div>
    </CenteredNicePaper>
  );
};

export default Experiment;
