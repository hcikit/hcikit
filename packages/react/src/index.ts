import { PropsWithChildren, ReactElement, Component } from "react";
import { ControlFunctions } from "./core/Experiment.js";
import * as allTasks from "./tasks/index.js";

export * from "./tasks/index.js";
export * from "./components/index.js";

export { allTasks };

export { withGridItem } from "./GridLayout.js";

export { default as GridLayout } from "./GridLayout.js";
export { default } from "./core/Experiment.js";
export * from "./core/Experiment.js";
export {
  BasePersistence,
  StoragePersistence,
  FileSystemPersistence,
} from "./persistence/index.js";

type TaskProps = ControlFunctions;
type StaticProps = {
  layoutProps?: Record<string, any>;
};

export type Task<P = {}> = React.ComponentType<TaskProps & P> & StaticProps;

export interface FunctionTask<P = {}>
  extends React.FunctionComponent<P & ControlFunctions> {
  (
    props: PropsWithChildren<P & ControlFunctions>,
    context?: any
  ): ReactElement<any, any> | null;
}

export abstract class ClassTask<P = {}, S = {}, SS = any> extends Component<
  P & TaskProps,
  S,
  SS
> {}
