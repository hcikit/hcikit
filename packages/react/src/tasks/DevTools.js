import React, { useEffect, useState } from "react";
import { IconButton, Slider, Card } from "@material-ui/core";
import {
  FastForward,
  SkipNext,
  SkipPrevious,
  FastRewind
} from "@material-ui/icons";
import styled from "styled-components";
import {
  taskNumberToIndex,
  indexToTaskNumber,
  getPropsFor,
  __INDEX__,
  getLeafIndex,
  getTotalTasks
} from "@hcikit/workflow";
import { withRawConfiguration } from "../core/withRawConfiguration";

import PropTypes from "prop-types";

const StyledCard = styled(Card)`
  display: inline-block;
  position: fixed;
`;
const Controls = styled.div`
  display: flex;
`;
const StyledSlider = styled(Slider)`
  margin-left: 10px;
  margin-right: 10px;
`;

export const DevTools = ({ taskComplete, configuration, setWorkflowIndex }) => {
  let [isDragging, setIsDragging] = useState(false);
  let [relativePosition, setRelativePosition] = useState();

  let [position, setPosition] = useState({ top: 0, left: 0 });

  useEffect(() => {
    function handleMouseMove(e) {
      if (isDragging) {
        // TODO: use bottom right instead.
        setPosition({
          top: e.pageY - relativePosition.y,
          left: e.pageX - relativePosition.x
        });
      }
    }

    function handleMouseUp() {
      setIsDragging(false);
    }

    window.addEventListener("mousemove", handleMouseMove);
    window.addEventListener("mouseup", handleMouseUp);

    return () => {
      window.removeEventListener("mouseup", handleMouseUp);
      window.removeEventListener("mousemove", handleMouseMove);
    };
  });

  // TODO: set the marks to the top level sections

  if (process.env.NODE_ENV === "production") {
    return null;
  }

  const topLevelTasks = configuration.children.map((_, i) => ({
    value: indexToTaskNumber(getLeafIndex([i], configuration), configuration),
    label: getPropsFor([i], configuration).task
  }));

  return (
    <StyledCard
      onMouseDown={e => {
        setIsDragging(true);
        setRelativePosition({
          x: e.pageX - position.left,
          y: e.pageY - position.top
        });
      }}
      style={position}
    >
      <Controls>
        <IconButton>
          <SkipPrevious />
        </IconButton>
        <IconButton>
          <FastRewind />
        </IconButton>
        <IconButton onClick={taskComplete}>
          <FastForward />
        </IconButton>
        <IconButton>
          <SkipNext />
        </IconButton>
      </Controls>
      <StyledSlider
        step={1}
        valueLabelDisplay="auto"
        value={indexToTaskNumber(configuration[__INDEX__], configuration)}
        marks={topLevelTasks}
        min={0}
        max={getTotalTasks(configuration)}
        onChange={(_, value) => {
          setWorkflowIndex(taskNumberToIndex(value, configuration));
        }}
      />
    </StyledCard>
  );
};

DevTools.propTypes = {
  configuration: PropTypes.object,
  setWorkflowIndex: PropTypes.func,
  taskComplete: PropTypes.func
};

export default withRawConfiguration(DevTools);
