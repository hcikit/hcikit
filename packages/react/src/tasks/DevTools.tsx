import React, { useEffect, useState } from "react";
import { IconButton, Slider, Card } from "@mui/material";
import FastForward from "@mui/icons-material/FastForward.js";
import SkipNext from "@mui/icons-material/SkipNext.js";
import SkipPreviousIcon from "@mui/icons-material/SkipPrevious.js";
import SkipPrevious from "@mui/icons-material/SkipPrevious.js";
import FastRewind from "@mui/icons-material/FastRewind.js";

import {
  taskNumberToIndex,
  indexToTaskNumber,
  getPropsFor,
  getLeafIndex,
  getTotalTasks,
  getCurrentIndex,
} from "@hcikit/workflow";

import _styled from "@emotion/styled";

const styled: typeof _styled =
  ((_styled as unknown as { default: typeof _styled })
    .default as typeof _styled) || _styled;

// TODO: making all of material ui icons a peer dependency instead of a dependency seems silly when they're probably just svgs abnd not dependent on the rest of material ui.

import { useConfiguration, useExperiment } from "../core/Experiment.js";

const StyledCard = styled(Card)`
  display: inline-block;
  position: fixed;
`;
const Controls = styled.div`
  display: flex;
`;
const StyledSlider = styled(Slider)`
  margin-left: 10px;
  margin-right: 10px;
`;

/**
 * TODO:
 * move to the bottom right or somewhere instead.
 * find a better way to represent labels, maybe a way to expand it?
 * add other things like a reset session or soemthing
 */

export const DevTools: React.FunctionComponent = () => {
  const configuration = useConfiguration();
  const experiment = useExperiment();

  const [isDragging, setIsDragging] = useState(false);
  const [relativePosition, setRelativePosition] = useState({ x: 0, y: 0 });

  const [position, setPosition] = useState({ top: 0, left: 0 });

  // TODO: I broke dragging
  useEffect(() => {
    function handleMouseMove(e: MouseEvent) {
      if (isDragging) {
        // TODO: use bottom right instead.
        setPosition({
          top: e.pageY - relativePosition.y,
          left: e.pageX - relativePosition.x,
        });
      }
    }

    function handleMouseUp() {
      setIsDragging(false);
    }

    window.addEventListener("mousemove", handleMouseMove);
    window.addEventListener("mouseup", handleMouseUp);

    return () => {
      window.removeEventListener("mouseup", handleMouseUp);
      window.removeEventListener("mousemove", handleMouseMove);
    };
    // TODO: I've done something wrong here. Like quite wrong.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  // TODO: set the marks to the top level sections

  if (process.env.NODE_ENV === "production") {
    return null;
  }

  let topLevelTasks: Array<{ value: number; label: string | undefined }> = [];

  if (configuration.children) {
    topLevelTasks = configuration.children.map((_: unknown, i: number) => ({
      value: indexToTaskNumber(configuration, getLeafIndex(configuration, [i])),
      label: getPropsFor(configuration, [i]).task,
    }));
  }

  return (
    <StyledCard
      onMouseDown={(e: React.MouseEvent) => {
        setIsDragging(true);
        setRelativePosition({
          x: e.pageX - position.left,
          y: e.pageY - position.top,
        });
      }}
      style={position}
    >
      <Controls>
        <IconButton>
          {/* @ts-ignore */}
          <SkipPrevious.default />
        </IconButton>
        <IconButton>
          {/* @ts-ignore */}
          <FastRewind.default />
        </IconButton>
        <IconButton onClick={() => experiment.advance()}>
          {/* @ts-ignore */}
          <FastForward.default />
        </IconButton>
        <IconButton size="large">
          {/* @ts-ignore */}
          <SkipNext.default />
        </IconButton>
      </Controls>
      <StyledSlider
        step={1}
        valueLabelDisplay="auto"
        value={indexToTaskNumber(configuration, getCurrentIndex(configuration))}
        marks={topLevelTasks}
        min={0}
        max={getTotalTasks(configuration)}
        onChange={(_: Event, value: number | Array<number>) => {
          experiment.advance(taskNumberToIndex(configuration, value as number));
        }}
      />
    </StyledCard>
  );
};

export default DevTools;
