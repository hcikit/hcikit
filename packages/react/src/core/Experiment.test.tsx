import React, { useState, useEffect, useRef } from "react";
import Experiment, {
  saveStateToSessionStorage,
  useConfig,
  useExperiment,
} from "./Experiment";
import DevTools from "../tasks/DevTools";
import DisplayText from "../tasks/DisplayTextTask";
import { render, screen, cleanup } from "@testing-library/react";
import { Configuration, Log } from "@hcikit/workflow";
import util from "util";
import userEvent from "@testing-library/user-event";

const config = {
  tasks: ["AuxTask"],
  children: [
    {
      text: "button task 1",
      task: "ButtonTask",
    },
    {
      text: "button task 2",
      task: "ButtonTask",
    },
  ],
};

const RenderCounter: React.FunctionComponent<{
  numRendersBeforeContinue: number;
  // eslint-disable-next-line react/prop-types
}> = ({ numRendersBeforeContinue = 2 }) => {
  const experiment = useExperiment();
  const renders = useRef(0);

  useEffect(() => {
    renders.current++;

    if (renders.current >= numRendersBeforeContinue) {
      experiment.advance();
    }
  });

  return <span data-testid="render-counter" />; //{renders}</span>;; // <span data-testid="renders">{renders}</span>;
};

const LogOnClick: React.FunctionComponent = () => {
  const experiment = useExperiment();
  return (
    <button
      onClick={() => {
        experiment.log({ type: "log" });
      }}
    >
      Button
    </button>
  );
};

const InfiniteRenderer: React.FunctionComponent = () => {
  const experiment = useExperiment();
  experiment.log({ type: "log" });
  return <button>Log</button>;
};

/* eslint-disable react/prop-types */
const ButtonTask: React.FunctionComponent<{
  text: string;
  logs: Array<Log>;
  configVal: string;
}> = ({ text, logs, configVal = "hello" }) => {
  const configuration = useConfig();
  const { log, modifyConfig, advance } = useExperiment();
  return (
    <>
      <button
        onClick={() => {
          log({ hello: "world", type: "log" });
          advance();
        }}
      >
        {text}
      </button>
      <p className="logs" data-testid="logs">
        {logs}
      </p>
      <p>{configVal}</p>
      <span onClick={() => modifyConfig({ configVal: "world" })}>
        Modify Config
      </span>
      <p>{configuration.children?.[1].foo as string}</p>
      <span onClick={() => modifyConfig({ foo: "bar" }, [1])}>at index</span>
    </>
  );
};

/* eslint-disable react/prop-types */
const MultiTask: React.FunctionComponent<{ times: number }> = ({ times }) => {
  const [timesClicked, setTimesClicked] = useState(0);
  const experiment = useExperiment();
  return (
    <button
      onClick={() => {
        setTimesClicked(timesClicked + 1);
        if (timesClicked + 1 >= times) {
          experiment.advance();
        }
      }}
    >
      {timesClicked}
    </button>
  );
};

const AuxTask = () => <p>Aux Task.</p>;

describe("Experiment", () => {
  it("renders the proper task", () => {
    render(
      <Experiment
        loadState={null}
        saveState={null}
        tasks={{ ButtonTask, AuxTask }}
        configuration={{ ...config }}
      />
    );

    screen.getByText(config.children[0].text);
  });

  it("advances to the next task", () => {
    render(
      <Experiment
        loadState={null}
        saveState={null}
        tasks={{ ButtonTask, AuxTask }}
        configuration={{ ...config }}
      />
    );

    screen.getByText("button task 1").click();
    screen.getByText(config.children[1].text);
  });

  it("advances to an arbitrary task", () => {
    render(
      <Experiment
        loadState={null}
        saveState={null}
        tasks={{ ButtonTask, AuxTask }}
        configuration={{ ...config }}
      />
    );

    fail();
  });

  it("works with empty tasks array", () => {
    const config = {
      content: "Hello",

      children: [
        {
          task: "DisplayText",
        },
      ],
    };

    render(
      <Experiment
        tasks={{ DisplayText }}
        loadState={null}
        saveState={null}
        configuration={{ ...config }}
      />
    );
    screen.getByText(config.content);
  });

  it("continues to the end", () => {
    render(
      <Experiment
        tasks={{ ButtonTask, AuxTask }}
        loadState={null}
        saveState={null}
        configuration={{ ...config }}
      />
    );

    screen.getByText("button task 1").click();
    screen.getByText("button task 2").click();
    screen.getByText("You've completed the experiment!");
  });

  it("tasks does not get bigger", () => {
    const Extender: React.FunctionComponent<{ tasks: Array<string> }> = ({
      tasks,
    }) => {
      const experiment = useExperiment();
      expect(tasks).toEqual(["Blank"]);

      return <button onClick={() => experiment.advance()}>Extender</button>;
    };

    const Blank = () => null;

    const config = {
      tasks: [],
      content: "Hello",
      children: [
        {
          tasks: ["Blank"],

          children: [
            { task: "Extender" },
            { task: "Extender" },
            { task: "Extender" },
            { task: "Extender" },
          ],
        },
      ],
    };

    render(
      <Experiment
        tasks={{ Extender, DisplayText, Blank }}
        loadState={null}
        saveState={null}
        configuration={{ ...config }}
      />
    );
    screen.getByText("Extender").click();
    screen.getByText("Extender").click();
    screen.getByText("Extender").click();
  });

  // this test is broken because log causes a state change in the infiniterenderer and this means they both render at once because of the infinite

  xit("logs definitely don't cause a re-render", () => {
    // I added this test because I had a log() statement that occurred on render and it caused an infinite loop.
    const config = {
      tasks: ["DisplayText"],
      content: "Hello",
      children: [
        { tasks: ["DisplayText", "RenderCounter"], task: "InfiniteRenderer" },
      ],
    };

    render(
      <Experiment
        tasks={{ InfiniteRenderer, DevTools, DisplayText, RenderCounter }}
        loadState={null}
        saveState={null}
        configuration={{ ...config }}
      />
    );

    screen.getByTestId("render-counter");
  });

  it("logs don't cause a re-render", () => {
    const config = {
      children: [
        {
          tasks: ["LogOnClick", "RenderCounter"],
        },
      ],
    };

    render(
      <Experiment
        tasks={{ RenderCounter, LogOnClick }}
        loadState={null}
        saveState={null}
        configuration={{ ...config }}
      />
    );

    screen.getByText("Button").click();
    screen.getByText("Button").click();
    screen.getByText("Button").click();
    screen.getByText("Button").click();
    screen.getByText("Button").click();
    screen.getByText("Button").click();
    screen.getByTestId("render-counter");
  });

  it("throws errors for unregistered tasks", () => {
    const consoleError = jest.spyOn(console, "error").mockImplementation(() => {
      //do nothing
    });

    const config = {
      children: [
        {
          times: 2,
          task: "MultiTask",
        },
      ],
    };

    expect(() => {
      render(
        <Experiment
          loadState={null}
          saveState={null}
          configuration={{ ...config }}
          tasks={{}}
        />
      );
    }).toThrow(Error);

    expect(consoleError).toHaveBeenCalled();
  });

  describe("uses sessionStorage properly", () => {
    it("loads from empty localStorage", () => {
      render(
        <Experiment
          tasks={{ ButtonTask, AuxTask }}
          configuration={{ ...config }}
        />
      );

      screen.getByText("button task 1").click();

      saveStateToSessionStorage.flush();

      cleanup();
      render(
        <Experiment
          tasks={{ ButtonTask, AuxTask }}
          configuration={{ ...config }}
        />
      );
      screen.getByText(config.children[1].text);
    });

    it("doesn't load when no task is given", () => {
      render(
        <Experiment
          tasks={{ ButtonTask, AuxTask }}
          configuration={{ ...config }}
        />
      );

      screen.getByText("button task 1").click();
      saveStateToSessionStorage.flush();

      cleanup();

      render(
        <Experiment
          loadState={null}
          saveState={null}
          tasks={{ ButtonTask, AuxTask }}
          configuration={{ ...config }}
        />
      );
      screen.getByText("button task 1");
    });

    afterEach(() => {
      window.sessionStorage.clear();
    });
  });

  describe("remounting", () => {
    it("remounts when forced remount", () => {
      const config = {
        children: [
          {
            times: 2,
            task: "MultiTask",
          },
          {
            times: 4,
            task: "MultiTask",
          },
        ],
      };

      render(
        <Experiment
          loadState={null}
          saveState={null}
          forceRemountEveryTask
          tasks={{ MultiTask }}
          configuration={{ ...config }}
        />
      );

      screen.getByText("0").click();
      screen.getByText("1").click();
      screen.getByText("0").click();
      screen.getByText("1").click();
      screen.getByText("2").click();
    });

    it("won't remount when not forced", () => {
      const config = {
        children: [
          {
            times: 2,
            task: "MultiTask",
          },
          {
            times: 4,
            task: "MultiTask",
          },
        ],
      };

      render(
        <Experiment
          loadState={null}
          saveState={null}
          forceRemountEveryTask={false}
          tasks={{ MultiTask }}
          configuration={{ ...config }}
        />
      );
      screen.getByText("0").click();
      screen.getByText("1").click();
      screen.getByText("2").click();
      screen.getByText("3").click();
    });
  });

  fit("logs properly", () => {
    var endConfiguration: Configuration = {};

    let Logger = () => {
      const { log, modifyConfig } = useExperiment();
      const [logValue, setLogValue] = useState("");
      return (
        <div>
          <input
            data-testid="log-value"
            type="text"
            value={logValue}
            onChange={(e) => setLogValue(e.target.value)}
          />
          <button onClick={() => log({ logValue, type: "log" })}>
            log as object
          </button>
          <button onClick={() => modifyConfig({ newConfig: "help" })}>
            modify config
          </button>
        </div>
      );
    };

    let LogOutput = () => {
      const config = useConfig();
      endConfiguration = config;

      return null;
    };

    let configuration: Configuration = {
      tasks: ["Logger", "LogOutput"],
      children: [
        { task: "ButtonTask", text: "button" },
        { task: "ButtonTask", text: "button" },
      ],
    };
    // do nothing
    let oldDate = Date.now;
    let i = 0;
    Date.now = () => i++;

    // TODO: monkeypatch date.now

    render(
      <Experiment
        tasks={{ Logger, LogOutput, ButtonTask }}
        configuration={configuration}
      />
    );

    userEvent.type(screen.getByTestId("log-value"), "logObject");

    screen.getByText("button").click();
    screen.getByText("modify config").click();

    screen.getByText("log as object").click();

    // TODO: I need to add the start and end logs in.
    screen.getByText("button").click();

    console.log(util.inspect(endConfiguration, undefined, null));

    if (endConfiguration?.children) {
      for (let child of endConfiguration.children) {
        console.log(child);
      }
    }
    // TODO: Check that the config modification gets logged too

    Date.now = oldDate;
  });

  it("doesn't pass logs to components", () => {
    render(
      <Experiment
        loadState={null}
        saveState={null}
        tasks={{ ButtonTask, AuxTask }}
        configuration={{ ...config }}
      />
    );

    expect(screen.getByTestId("logs")).toBeEmptyDOMElement();
    screen.getByText("button task 1").click();
    expect(screen.getByTestId("logs")).toBeEmptyDOMElement();
  });

  it("modifies config properly", () => {
    render(
      <Experiment
        loadState={null}
        saveState={null}
        tasks={{ ButtonTask, AuxTask }}
        configuration={{ ...config }}
      />
    );

    screen.getByText("hello");
    screen.getByText("Modify Config").click();
    screen.getByText("world");

    screen.getByText("at index").click();
    screen.getByText("bar");
  });
  // TODO: modify config at depth works too.
  it("modifies config at depth properly", () => {});
});
