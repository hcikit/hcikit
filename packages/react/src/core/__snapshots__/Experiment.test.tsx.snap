// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorHandler handles errors properly 1`] = `
<DocumentFragment>
  <div
    id="grid-layout"
    style="display: grid; width: 100%; height: 100%; grid-template-columns: 1fr; grid-template-rows: min-content 1fr min-content; grid-template-areas: \\"header\\"
      \\"task\\"
      \\"footer\\";"
  >
    <button>
      button
    </button>
    <p
      class="logs"
      data-testid="logs"
    />
    <p>
      hello
    </p>
    <span>
      Modify Config
    </span>
    <p />
    <span>
      at index
    </span>
  </div>
</DocumentFragment>
`;
